RecoCTPPS/PixelLocal/src/RPixDetClusterizer.cc

```diff
@@ -0,0 +1,166 @@
+#include <iostream>
+#include "FWCore/Utilities/interface/Exception.h"
+
+#include "RecoCTPPS/PixelLocal/interface/RPixDetClusterizer.h"
+
+namespace {
+  constexpr int max16bits = 65535;
+  constexpr int maxCol = 155;
+  constexpr int maxRow = 159;
+  constexpr double highRangeCal = 1800.;
+  constexpr double lowRangeCal = 260.;
+}
+
+
+RPixDetClusterizer::RPixDetClusterizer(edm::ParameterSet const& conf):
+  params_(conf), SeedVector_(0)
+{
+  verbosity_ = conf.getUntrackedParameter<int>("RPixVerbosity");
+  SeedADCThreshold_ = conf.getParameter<int>("SeedADCThreshold");
+  ADCThreshold_ = conf.getParameter<int>("ADCThreshold");
+  ElectronADCGain_ = conf.getParameter<double>("ElectronADCGain");
+  VcaltoElectronGain_ = conf.getParameter<int>("VCaltoElectronGain");
+  VcaltoElectronOffset_ = conf.getParameter<int>("VCaltoElectronOffset");
+  doSingleCalibration_ = conf.getParameter<bool>("doSingleCalibration");
+
+}
+
+RPixDetClusterizer::~RPixDetClusterizer(){}
+
+void RPixDetClusterizer::buildClusters(unsigned int detId, const std::vector<CTPPSPixelDigi> &digi, std::vector<CTPPSPixelCluster> &clusters, const CTPPSPixelGainCalibrations * pcalibrations,const CTPPSPixelAnalysisMask* maskera)
+{
+
+  std::map<uint32_t, CTPPSPixelROCAnalysisMask> mask = maskera->analysisMask;
+  std::map<uint32_t, CTPPSPixelROCAnalysisMask>::iterator mask_it = mask.find(detId); 
+
+  std::set<std::pair<unsigned char, unsigned char> > maskedPixels;
+  if( mask_it != mask.end()) maskedPixels = mask_it->second.maskedPixels;
+
+  if(verbosity_) edm::LogInfo("RPixDetClusterizer")<<detId<<" received digi.size()="<<digi.size();
+
+// creating a set of CTPPSPixelDigi's and fill it
+
+  rpix_digi_set_.clear();
+  rpix_digi_set_.insert(digi.begin(),digi.end());
+
+// calibrate digis here
+  calib_rpix_digi_set_.clear();
+
+  for( std::set<CTPPSPixelDigi>::iterator RPdit = rpix_digi_set_.begin(); RPdit != rpix_digi_set_.end();++RPdit){
+
+    if((*RPdit).row() > maxRow || (*RPdit).column() > maxCol)
+      throw cms::Exception("CTPPSDigiOutofRange") 
+	<< " row = " << (*RPdit).row() << "  column = "<< (*RPdit).column();
+
+    unsigned char row = (*RPdit).row();
+    unsigned char column = (*RPdit).column();
+    unsigned short adc = (*RPdit).adc();
+    unsigned short electrons = 0;
+    std::pair<unsigned char, unsigned char> pixel = std::make_pair(row,column);
+
+// check if pixel is noisy or dead (i.e. in the mask)
+    const bool is_in = maskedPixels.find(pixel) != maskedPixels.end();
+    if(!is_in){
+//calibrate digi and store the new ones
+      electrons = calibrate(detId,adc,row,column,pcalibrations);
+      RPixCalibDigi calibDigi(row,column,adc,electrons);
+      calib_rpix_digi_set_.insert(calibDigi);
+    }
+  }
+  if(verbosity_) edm::LogInfo("RPixDetClusterizer")<<" RPix set size = "<<calib_rpix_digi_set_.size();
+
+  SeedVector_.clear();
+
+// clusterizing and storing the seeds
+  for (const auto &rpcd : calib_rpix_digi_set_){
```

Comment:
Probably the make_cluster(...) method can be optimized further... But ok, let it as such for moment

Anyhow, at least this loop can be removed even with the current make_cluster(...):
- Move SeedVector_.clear(); before L49
- Add if(calibDigi.electrons() > SeedADCThreshold_*ElectronADCGain_) SeedVector_.push_back(calibDigi); after L67
- Remove L74 to L79